packages/client/hmi-client/src/components/responsive-matrix/matrix.vue

```diff
@@ -0,0 +1,902 @@
+
+		// ///////////////////////////////////////////////////////////////////////////////
+
+		// initialize screen height/width
+		this.screenHeight = matrix.offsetHeight;
+		this.screenWidth = matrix.offsetWidth;
+
+		// start the resize observer
+		this.resizeObserver = new ResizeObserver((entries) => {
+			this.screenHeight = entries[0].contentRect.height;
+			this.screenWidth = entries[0].contentRect.width;
+		});
+		this.resizeObserver.observe(matrixContainer);
+
+		// start event listeners
+		window.addEventListener('keydown', this.handleKey);
+		window.addEventListener('keyup', this.handleKey);
+		matrixContainer.addEventListener('wheel', this.handleScroll);
+
+		// ///////////////////////////////////////////////////////////////////////////////
+
+		// initialize pixi.js
+		const app = new Application({
+			resizeTo: matrixContainer,
+			backgroundColor: this.backgroundColor
+		});
+		matrix.appendChild(app.view as unknown as Node);
+
+		// create viewport
+		const screenHeight = this.screenHeight;
+		const screenWidth = this.screenWidth;
+		const cellAspectRatio = 1 || this.microRowArray.length / this.microColArray.length;
+		const quadAspectRatio = this.screenAspectRatio * cellAspectRatio; // height / width
+		const vertexPosition =
+			quadAspectRatio > 1 // is height long?
+				? [
+						-1 / quadAspectRatio,
+						-1, // height long
+						1 / quadAspectRatio,
+						-1, // x, y
+						1 / quadAspectRatio,
+						1,
+						-1 / quadAspectRatio,
+						1
+				  ]
+				: [
+						-1,
+						-1 * quadAspectRatio, // width long
+						1,
+						-1 * quadAspectRatio, // x, y
+						1,
+						1 * quadAspectRatio,
+						-1,
+						1 * quadAspectRatio
+				  ];
+		this.worldHeight = vertexPosition[5] - vertexPosition[3];
+		this.worldWidth = vertexPosition[2] - vertexPosition[0];
+		(this as any).$viewport = new Viewport({
+			screenWidth: this.screenWidth,
+			screenHeight: this.screenHeight,
+			worldWidth: this.worldWidth,
+			worldHeight: this.worldHeight,
+			divWheel: matrix as HTMLElement,
+			interaction: app.renderer.plugins.interaction // the interaction module is important for wheel to work properly when renderer.view is placed or scaled
+		});
+
+		// add the viewport to the stage
+		app.stage.addChild((this as any).$viewport as unknown as DisplayObject);
+
+		// activate camera interaction plugins
+		(this as any).$viewport.drag().pinch().wheel();
+		(this as any).$viewport.pause = true;
+
+		// clamp camera
+		(this as any).$viewport.clampZoom({
+			maxWidth: this.worldWidth * 2,
+			maxHeight: this.worldHeight * 2
+		});
+
+		// build quad
+		const geometry = new Geometry()
+			.addAttribute('aVertexPosition', vertexPosition, 2) // the size of the attribute
+			.addAttribute(
+				'aUvs',
+				[
+					0,
+					0, // u, v
+					1,
+					0,
+					1,
+					1,
+					0,
+					1
+				], // u, v
+				2
+			) // the size of the attribute
+			.addIndex([0, 1, 2, 0, 2, 3]);
```

Comment:
Out of curiosity, what does addIndex do here?

Reply:
its specifying the points making up the 2 triangles forming the quad. so what this is saying here is the first triangle are attributes of index 0, 1, 2, and the second triangle is formed by attributes of index 0, 2, 3

Reply:
I see, I thought it would've been a triangle-strip or something, e.g. not size=6.